### PR TITLE
Refactor error handling for implicit actions

### LINKED ISSUE #32694: Improve FK violation error messages when running basic model actions
Epic: #32320

See: 
- https://metaboat.slack.com/archives/C057T1QTB3L/p1690451722933519
- https://github.com/metabase/metabase/pull/32556#pullrequestreview-1545650897
- https://github.com/metabase/metabase/pull/32556#issuecomment-1650353266
- https://metaboat.slack.com/archives/C057LE0Q4PQ/p1690295017296989?thread_ts=1690208614.530869&cid=C057LE0Q4PQ

### PR DESCRIPTION
**Pull request by qnkhuat**
Closes https://github.com/metabase/metabase/issues/32694

We're working on making actions error more user-friendly ([thread](https://metaboat.slack.com/archives/C057T1QTB3L/p1690462817579789?thread_ts=1690451722.933519&cid=C057T1QTB3L))

In this PR I refactored the error handling flow with 2 main goals:
- Making defining SQL error message parser easier and unifying the shape of error data
- Reduce the amount of custom try-catch in the call stack

Additional to that:
- we add tests around to cover all error cases
- add a parser for h2 which didn't exist before
- updated the error message to be more friendly([thread](https://metaboat.slack.com/archives/C057T1QTB3L/p1692023106337609?thread_ts=1690451722.933519&cid=C057T1QTB3L))

```
- Violate not null
  - Summarize : Name, <field-X>, and Email must have values.
  - Per column : You must provide a value.
- Violate unique constraints
  - Summarize: URL and Short code already exist.
  - Per column: This URL value already exists.
- Violate incorrect type
  - Summarize: Some of your values aren’t of the correct type for the database.
  - Per column: This value needs to be <Text | a Number | a Date>
- Violate FK constraint
  - Delete
   - Summarize: Other tables rely on this record so it cannot be deleted.
   - Per column: N/A I think?
  - Create
   - Summarize: Unable to create a new record
   - Per column: This Product ID does not exist
  - Update
   - Summarize: Unable to update the record
   - Per column: This Product ID does not exist
```

### CODE REVIEW

File: src/metabase/actions/execution.clj [hunk 1]
```diff
@@ -159,11 +156,8 @@
 
                   (= implicit-action :row/update)
                   (assoc :update-row row-parameters))]
-    (try
-      (binding [qp.perms/*card-id* (:model_id action)]
-        (actions/perform-action! implicit-action arg-map))
-      (catch Exception e
-        (handle-action-execution-error e)))))
+    (binding [qp.perms/*card-id* (:model_id action)]
+      (actions/perform-action! implicit-action arg-map))))
```
Comment (qnkhuat) [inline src/metabase/actions/execution.clj:160]: I think we need a nice job catching and rethrowing errors inside `perform-action!`, there is no need to do another try catch here.

File: src/metabase/driver/mysql/actions.clj [hunk 2]
```diff
@@ -52,67 +55,55 @@
...
+                                          "(conn=21) Cannot delete or update a parent row: a foreign key constraint fails (`action-error-handling`.`user`, CONSTRAINT `user_group-id_group_-159406530` FOREIGN KEY (`group-id`) REFERENCES `group` (`id`))")
+  (sql-jdbc.actions/maybe-parse-sql-error :mysql actions.error/violate-unique-constraint {:id 3}
+                                          "(conn=10) Duplicate entry 'ID' for key 'string_pk.PRIMARY'")
+  (sql-jdbc.actions/maybe-parse-sql-error :mysql actions.error/violate-not-null-constraint nil
+                                          "Column 'f1' cannot be null")
+  (sql-jdbc.actions/maybe-parse-sql-error :mysql actions.error/incorrect-value-type nil
+                                          "(conn=183) Incorrect integer value: 'STRING' for column `table`.`id` at row 1")
   nil)
 
-(defmethod sql-jdbc.actions/parse-sql-error :mysql
```
Comment (qnkhuat) [inline src/metabase/driver/mysql/actions.clj:108]: removing `sql-jdbc.actions/parse-sql-error` was my original motivation, I think having a list of functions for each driver like this is hard to know which drivers support which parsing error function.

File: src/metabase/driver/sql_jdbc/actions.clj [hunk 3]
```diff
@@ -23,41 +24,68 @@
...
-  "Returns errors in a way that indicates which column had the problem. Can be used to highlight errors in forms."
+(defmethod maybe-parse-sql-error :default
+  [_driver _error-type _database _e]
+  nil)
+
+(defn- parse-sql-error
   [driver database e]
-  (let [default-method? (= (get-method parse-sql-error driver)
-                           (get-method parse-sql-error :default))]
-    (if default-method?
```
Comment (qnkhuat) [inline src/metabase/driver/sql_jdbc/actions.clj:42]: this logic was also the motivation for the refactor, it's really hard to know what will be parsed upstream and what's being processed.

**Review by metamben — COMMENTED** [review]
I like the direction this is going.

File: src/metabase/driver/mysql/actions.clj [hunk 4]
```diff
@@ -52,67 +55,53 @@
...
-                        :column     col}))))))
+(defmethod sql-jdbc.actions/maybe-parse-sql-error [:mysql actions.error/violate-unique-constraint]
+  [_driver error-type database error-message]
+  (when-let [[_match table constraint]
+             (re-find #"Duplicate entry '.+' for key '(.+)\.(.+)'" error-message)]
+    (let [constraint (remove-backticks constraint)
+          table      (remove-backticks table)
+          columns    (constraint->column-names database table constraint)]
+      {:type    error-type
+       :message (tru "Value for column(s) {0} is duplicated" (str/join ", " columns))
```
[the remark this answers is not in the document]
Comment (tsmacdonald) [reply src/metabase/driver/mysql/actions.clj:66]: +1, we can do it. Look at https://github.com/metabase/metabase/blob/34d895a30dac2832d38a48ef339da7e7ad5cb4ac/test/metabase/util/i18n_test.clj#L112 for inspiration

File: src/metabase/driver/sql_jdbc/actions.clj [hunk 5]
```diff
@@ -23,41 +24,68 @@
...
-            (merge {:exception-data (ex-data e)
-                    :status-code status-code}
-                   more-info)
-            e)))
+  (try
+   (some #(maybe-parse-sql-error driver % database (ex-message e))
+          [actions.error/violate-unique-constraint
+           actions.error/violate-foreign-key-constraint
+           actions.error/violate-not-null-constraint
+           actions.error/incorrect-value-type])
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/driver/sql_jdbc/actions.clj:58]: > What if a driver has some other error type, shouldn't it be able to return it?

I've changed this static list to a method to get all error types given a driver. so any custom driver error parser should be able to return.

> What if a driver can easily tell what error type the message is about (e.g., the message starts with an error code)

Notice the try catch in `do-with-auto-parse-sql-error` catch the SQLException class, so if there are error that driver can recognize, just throw a normal exception, it'll not get parsed.

File: src/metabase/driver/postgres/actions.clj [hunk 6]
```diff
@@ -2,22 +2,26 @@
...
-             (re-find #"ERROR:\s+(\w+) value in column \"([^\"]+)\" violates not-null constraint" error-message)]
-    [{:message (tru "violates not-null constraint")
-      :column column}]))
+(defmethod sql-jdbc.actions/maybe-parse-sql-error [:postgres actions.error/violate-not-null-constraint]
+  [_driver error-type _database error-message]
+  (when-let [[_ _value column _table]
+             (re-find #"ERROR:\s+(\w+) value in column \"([^\"]+)\" of relation \"([^\"]+)\" violates not-null constraint"  error-message)]
+    {:type    error-type
+     :message (tru "Value for column {0} must be not null" column)
+     :errors  {column (tru "The value must be not null")}}))
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/driver/postgres/actions.clj:20]: Could you elaborate? I don't understand what default implementation you want here.

> I expect this map to be the same across almost all DB types.

yes, that's basically the ide aof `maybe-parse-sql-error`, we expect it either returns `nil`, or a map with these 3 keys.

File: src/metabase/actions/error.clj [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  "Error type for SQL not null constraint violation."
+  ::violate-not-null-constraint)
+
+(def violate-foreign-key-constraint
+  "Error type for SQL foreign key constraint violation."
+  ::violate-foreign-key-constraint)
+
+(def incorrect-value-type
+  "Error type for SQL incorrect value type."
+  ::incorrect-value-type)
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/actions/error.clj:17]: My motivation for defining var for these things are:
- we can provide docs
- easy to find all error types
- centralized and maybe in the future we can do sth like error hierarchy as we have in `metabase.query-processor.error-type`

I could remove this ns if you find it's not justified. normal keywords also work for me.

File: src/metabase/driver/sql_jdbc/actions.clj [hunk 8]
```diff
@@ -23,41 +24,72 @@
...
-
-(defn- catch-throw [e status-code & [more-info]]
-  (throw
-   (ex-info (ex-message e)
-            (merge {:exception-data (ex-data e)
-                    :status-code status-code}
-                   more-info)
-            e)))
+  (try
+   (some #(maybe-parse-sql-error driver % database (ex-message e)) (driver->available-error-parser driver))
```
Comment (calherries) [inline src/metabase/driver/sql_jdbc/actions.clj:61]: This could be simplified. `(methods maybe-parse-sql-error)` returns a dispatch table of driver and error-type to parser function. You're effectively looking up this table twice just to get the parser functions. Once to get the error-types for a driver, and then using those error-types, you get the methods. But you could just do it in one step, where you get the methods for a given driver and apply them directly.

e.g.

```clojure
(defn- parse-sql-error
  [driver database e]
  (let [parsers-for-driver (keep (fn [[[method-driver _error-type] method]]
                                   (when (= method-driver driver)
                                     method))
                                 (dissoc (methods maybe-parse-sql-error) :default))]
    (try
      (some #(% driver database (ex-message e)) parsers-for-driver)
      ;; Catch errors in parse-sql-error and log them so more errors in the future don't break the entire action.
      ;; We'll still get the original unparsed error message.
      (catch Throwable new-e
        (log/error new-e (trs "Error parsing SQL error message {0}: {1}" (pr-str (ex-message e)) (ex-message new-e)))
        nil))))
```

File: src/metabase/actions/error.clj (hunk 7 above)
[the remark this answers is not in the document]
Comment (calherries) [reply src/metabase/actions/error.clj:17]: It might be useful to add another layer of indirection and centralize the creation of the error maps currently returned by each implementation of `maybe-parse-sql-error`. For each error-type, we need to return the same format of data under the `:message` and `:errors` keys. We should only need to define the `:message` and error messages under `:errors` in one place for each error-type, not once for each combination of driver and error-type. Then the only thing you would need to define in each driver's maybe-parse-sql-error implementation is how to extract the required components from the DBs error message.

[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/actions/error.clj:17]: @calherries 
Initially, I went with that approach and I really hope we could.

but the problem is the error message from each driver does not provide the same level of granularity.

For example for incorrect type error.
Mysql error message: `Incorrect integer value: 'STRING' for column "table.id" at row 1`
Postgres error message: `invalid input syntax for type integer: \"S\"`

Notice the error message from Postgres does not contain which column it has type error. 
So if we want the message and errors to be the same for all drivers, it enforces the error message to be generic and requires as few components as possible.

That's why I changed to "error" per driver approach like this, it's extra effort, but it gives more flexibility.

File: src/metabase/driver/postgres/actions.clj [hunk 9]
```diff
@@ -27,38 +31,55 @@
...
+  :postgres actions.error/violate-unique-constraint {:id 47}
+  "Batch entry 0 UPDATE \"public\".\"group\" SET \"ranking\" = CAST(2 AS INTEGER) WHERE \"public\".\"group\".\"id\" = 1 was aborted: ERROR: duplicate key value violates unique constraint \"group_ranking_key\"\n  Detail: Key (ranking)=(2) already exists.  Call getNextException to see other errors in the batch.")
+
+ (sql-jdbc.actions/maybe-parse-sql-error
+  :postgres actions.error/violate-not-null-constraint nil
+  "ERROR: null value in column \"ranking\" of relation \"group\" violates not-null constraint\n  Detail: Failing row contains (57, admin, null).")
+
+ (sql-jdbc.actions/maybe-parse-sql-error
+  :postgres actions.error/incorrect-value-type nil
+  "Batch entry 0 UPDATE \"public\".\"group\" SET \"ranking\" = CAST('S' AS INTEGER) WHERE \"public\".\"group\".\"id\" = 1 was aborted: ERROR: invalid input syntax for type integer: \"S\"  Call getNextException to see other errors in the batch."))
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/driver/postgres/actions.clj:82]: I'll add more tests later, but I don't think we should have error strings as test cases, I want to go further and construct the actually query to throw such errors so that if message error changes between driver version, we can catch that and write more robust regex to cover those cases.

**Comment by qnkhuat** [comment]
Could anyone give another look at the overall shape? I'll polish the error message once we agree on the design.

**Review by calherries — APPROVED** [review]
Nice

File: src/metabase/driver/sql_jdbc/actions.clj (hunk 5 above)
[the remark this answers is not in the document]
Comment (metamben) [reply src/metabase/driver/sql_jdbc/actions.clj]: If that's the suggested way, we should probably document it. I expect most driver authors would want to profit from the existing framework as much as possible and might not think about doing everything from "scratch".

It's much easier to understand and maintain a set of functions with clear interfaces than throwing the exceptions of the right shape and info. If we go and document the exception "interface" then it's not a problem though.

File: src/metabase/driver/postgres/actions.clj (hunk 6 above)
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/driver/postgres/actions.clj]: I see, I think this comment cover that https://github.com/metabase/metabase/pull/32935#discussion_r1293142525

TLDR is that I initially did something similar to that, but error messages from each driver does not contain the same level of granularity so it's hard to generalize these.

File: src/metabase/driver/sql_jdbc/actions.clj (hunk 5 above)
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/driver/sql_jdbc/actions.clj]: https://github.com/metabase/metabase/blob/bcb605ec5bee47b3a9d7e40eb9e0159e18712d84/src/metabase/driver/sql_jdbc/actions.clj#L34

is the docstring here state that well?

File: src/metabase/driver/mysql/actions.clj [hunk 10]
```diff
@@ -52,67 +55,53 @@
...
+       :errors  (reduce (fn [acc col]
+                          (assoc acc col (tru "This column has unique constraint and this value is existed")))
+                        {}
+                        columns)})))
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/driver/mysql/actions.clj:76]: we always have a summarized message for that, I think the intention of these per-column errors are to aid users in fixing the wrong value.

**Comment by qnkhuat** [comment]
@calherries I've updated the error messages and tests could you please do another review? You can find the spec in the PR's description

File: src/metabase/driver/mysql/actions.clj [hunk 11]
```diff
@@ -52,66 +40,82 @@
...
+      :message (tru "{0} must have values." (str/capitalize column))
+      :errors  {column (tru "You must provide a value.")}})))
+
+(defmethod sql-jdbc.actions/maybe-parse-sql-error [:mysql actions.error/violate-unique-constraint]
+  [_driver error-type _database _action-type error-message]
+  (when-let [[_match fk]
+             (re-find #"Duplicate entry '.+' for key '(.+)'" error-message)]
+    (let [column (last (str/split fk #"\."))]
+      {:type    error-type
+       :message (tru "{0} already exist." (str/capitalize column))
```
Comment (calherries) [inline src/metabase/driver/mysql/actions.clj:69]: There's only one column, so this needs to be
```suggestion
       :message (tru "{0} already exists." (str/capitalize column))
```

File: src/metabase/util/string.clj [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+  ([parts]
+   (build-sentence parts :stop? true))
+  ([parts & {:keys [stop?]
+             :or   {stop? true}
+             :as _options}]
+   (when (seq parts)
+     (cond
+       (= (count parts) 1) (str (first parts) (when stop? \.))
+       (= (count parts) 2) (str (first parts) " " (deferred-tru "and")  " " (second parts) (when stop? \.))
+       :else               (str (first parts) ", " (build-sentence (rest parts)))))))
```
Comment (calherries) [inline src/metabase/util/string.clj:28]: This doesn't work for all languages, especially right-to-left ones. And I'm pretty sure not all languages will use commas and "and" in that way to create lists. I'm not sure what the solution is, but this isn't it. Maybe someone has attempted to solve the problem before and has some clues, especially on the FE

File: src/metabase/driver/postgres/actions.clj [hunk 13]
```diff
@@ -27,38 +23,75 @@
...
+     :message (tru "{0} must have values." (str/capitalize column))
+     :errors  {column (tru "You must provide a value.")}}))
+
+(defmethod sql-jdbc.actions/maybe-parse-sql-error [:postgres actions.error/violate-unique-constraint]
+  [_driver error-type database _action-type error-message]
+  (when-let [[_match constraint _value]
+             (re-find #"duplicate key value violates unique constraint \"([^\"]+)\"" error-message)]
+    (let [columns (constraint->column-names database constraint)]
+      {:type    error-type
+       :message (tru "{0} already exist." (u/build-sentence (map str/capitalize columns) :stop? false))
```
Comment (calherries) [inline src/metabase/driver/postgres/actions.clj:40]: This needs to be "exists" when there is one column. You can use `trun` instead of `tru` for that.

File: src/metabase/driver/h2/actions.clj [hunk 14]
```diff
@@ -21,3 +28,85 @@
...
+  (when-let [[_match _constraint-name _table column]
+             (re-find #"Referential integrity constraint violation: \"([^\:]+): [^.]+.\"\"([^\s]+)\"\" FOREIGN KEY\(\"\"([^\"]+)\"\"\)" error-message)]
+    (merge {:type error-type}
+           (case action-type
+             :row/create
+             {:message (tru "Unable to create a new record.")
+              :errors {column (tru "This {0} does not exist." (str/capitalize column))}}
+
+             :row/delete
+             {:message (tru "Other tables rely on this row so it cannot be updated/deleted.")
```
Comment (calherries) [inline src/metabase/driver/h2/actions.clj:84]: why updated?
```suggestion
             {:message (tru "Other tables rely on this row so it cannot be deleted.")
```

**Review by calherries — CHANGES_REQUESTED** [review]

File: src/metabase/driver/h2/actions.clj (hunk 14 above)
Comment (qnkhuat) [reply src/metabase/driver/h2/actions.clj]: ah it should be removed. updated!

File: src/metabase/driver/postgres/actions.clj (hunk 13 above)
Comment (qnkhuat) [reply src/metabase/driver/postgres/actions.clj]: that's a bit tricky to use. [ecfde21](https://github.com/metabase/metabase/pull/32935/commits/ecfde213405235d7af83110b062b67654b1e87cf)

File: src/metabase/driver/mysql/actions.clj (hunk 11 above)
Comment (qnkhuat) [reply src/metabase/driver/mysql/actions.clj]: good catch. Thanks

File: src/metabase/util/string.clj (hunk 12 above)
Comment (qnkhuat) [reply src/metabase/util/string.clj]: https://shopify.engineering/internationalization-i18n-best-practices-front-end-developers

I read this the other day and I don't think they have a solution for dynamically building sentences.

Also,
1. Solving the generic problem of pluralization shouldn't be the goal of this PR
2. We only support 29 languages, 2 of which are right-to-left, so I think our pluralization is far from perfect. not that it's an excuse, but it supports the 1st point.
3. As this is exacted to be its own function, I think once we decided to solve this problem, we can solve it once for all.
4. Started a discussion here https://metaboat.slack.com/archives/C010L1Z4F9S/p1692241389613209

**Comment by kamilmielnik** [comment]
> Closes #32935

It's this PR's own id. Closes #32694

File: src/metabase/driver/postgres/actions.clj [hunk 15]
```diff
@@ -27,38 +23,75 @@
...
+                 (re-find #"update or delete on table \"([^\"]+)\" violates foreign key constraint \"([^\"]+)\" on table \"([^\"]+)\"\n  Detail: Key \((.*?)\)=\((.*?)\) is still referenced from table \"([^\"]+)\"" error-message)]
+        (merge {:type error-type}
+               (case action-type
+                 :row/delete
+                 {:message (tru "Other tables rely on this row so it cannot be deleted.")
+                  :errors  {}}
+
+                 :row/update
+                 {:message (tru "Unable to update the record.")
+                  :errors  {column (tru "This {0} does not exist." (str/capitalize column))}})))
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/driver/postgres/actions.clj:58]: done!

[the remark this answers is not in the document]
Comment (kamilmielnik) [reply src/metabase/driver/postgres/actions.clj:58]: Works well, thanks!

File: test/metabase/api/action_test.clj [hunk 16]
```diff
@@ -633,3 +639,99 @@
...
+                              {update-user :action-id} {:type :implicit
+                                                        :kind "row/update"}]
+              (testing "violate not-null constraint"
+                (testing "when creating"
+                  (is (= {:message "Ranking must have values."
+                          :errors {:ranking "You must provide a value."}}
+                         (mt/user-http-request :rasta :post 400 (format "action/%d/execute" create-group)
+                                               {:parameters {"name" "admin" "ranking" nil}}))))
+                (testing "when updating"
+                  (is (= "Implicit parameters must be provided."
```
Comment (calherries) [inline test/metabase/api/action_test.clj:680]: Is this because there's no way to update a value to be null?

File: src/metabase/util/string.clj (hunk 12 above)
Comment (calherries) [reply src/metabase/util/string.clj]: Good points. Should we add a small FIXME at least explaining that this doesn't work for a lot of languages?

File: test/metabase/api/action_test.clj [hunk 17]
```diff
@@ -633,3 +639,99 @@
...
+                         (mt/user-http-request :rasta :post 400 (format "action/%d/execute" update-group)
+                                               {:parameters {"id" "admin" "ranking" nil}})))))
+
+              (testing "violate unique constraint"
+                (testing "when creating"
+                  (is (= {:message "Ranking already exists."
```
Comment (calherries) [inline test/metabase/api/action_test.clj:686]: We should probably have a test for the plural form, when there are multiple columns involved.

File: src/metabase/driver/h2/actions.clj [hunk 18]
```diff
@@ -29,6 +29,19 @@
...
+(defn- db-identifier->name
+  "Get the name of identifier from JDBC error message.
+  An identifier can contains quote and full schema, database, table , etc.
+  This formats so that we get  only the identifer name with quote removed.
+
+
+    (db-identifier->name \"PUBLIC.TABLE\" ) => \"TABLE\""
+  [s]
+  (-> s
+      (str/replace #"\"" "")
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/driver/h2/actions.clj:41]: sorry but this replace the quote, not backslash. or am I missing something?

[the remark this answers is not in the document]
Comment (calherries) [reply src/metabase/driver/h2/actions.clj:41]: Wait, is this the point of this line to unescape escaped quotes? If so explain with an example

[the remark this answers is not in the document]
Comment (calherries) [reply src/metabase/driver/h2/actions.clj:41]: Ah, I'm mistaken. Never mind

File: src/metabase/actions/error.clj [hunk 19]
```diff
@@ -0,0 +1,17 @@
+(ns metabase.actions.error)
```
Comment (kamilmielnik) [inline src/metabase/actions/error.clj:1]: _leaving comment in an unrelated place_

I tried to execute an action with parameters that would violate 2 different FK constraints, but I'm only getting error for one of them:

image

----

If I try to violate constraints separately, I correctly receive the detailed errors:

image

File: test/metabase/api/action_test.clj [hunk 20]
```diff
@@ -633,3 +639,99 @@
...
+    [["admin" 1]
+     ["user" 2]]]
+   ["user"
+    [{:field-name "name" :base-type :type/Text :not-null? true}
+     {:field-name "group-id" :base-type :type/Integer :fk "group" :not-null? true}]
+    [["crowberto" 1]
+     ["rasta"     2]
+     ["lucky"     1]]]])
+
+(deftest action-error-handling-test
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply test/metabase/api/action_test.clj:656]: no strong reason, we can refactor into 2 tests like that and can test the case `violate not-null constraint` when updating as well.

Will do tmr.

File: test/metabase/api/action_test.clj [hunk 21]
```diff
@@ -633,3 +639,99 @@
...
+                                                       :kind "row/delete"}]
+            (mt/with-actions [{_card-id :id}           {:dataset_query (mt/mbql-query user) :dataset true}
+                              {create-user :action-id} {:type :implicit
+                                                        :kind "row/create"}
+                              {update-user :action-id} {:type :implicit
+                                                        :kind "row/update"}]
+              (testing "violate not-null constraint"
+                (testing "when creating"
+                  (is (= {:message "Ranking must have values."
+                          :errors {:ranking "You must provide a value."}}
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply test/metabase/api/action_test.clj:676]: yea, `column-errors` might be better described that, but I have 2 counter points:
- errors are also used in the API to return `per param schema error`, so that's might be an existing pattern?
- our middleware have a special treatment for `:errors` though ([here](https://github.com/metabase/metabase/blob/ngoc-actions-error-handling-cleanup/src/metabase/server/middleware/exceptions.clj#L65)), so if we change to `column-errors` we'd need to treat it a bit different. Also this is an related [issue](https://github.com/metabase/metabase/issues/14683)

[the remark this answers is not in the document]
Comment (calherries) [reply test/metabase/api/action_test.clj:676]: Fair points, lets keep the `:errors` key. I wasn't aware that's how the middleware works.

File: test/metabase/api/action_test.clj (hunk 20 above)
[the remark this answers is not in the document]
Comment (qnkhuat) [reply test/metabase/api/action_test.clj:644]: did this in [cc95f07](https://github.com/metabase/metabase/pull/32935/commits/cc95f072c399c1165df95478a655e597974483ee)

File: src/metabase/util/string.clj (hunk 12 above)
Comment (qnkhuat) [reply src/metabase/util/string.clj]: I added that as a note in the function https://github.com/metabase/metabase/blob/ngoc-actions-error-handling-cleanup/src/metabase/util/string.clj

File: test/metabase/api/action_test.clj (hunk 17 above)
Comment (qnkhuat) [reply test/metabase/api/action_test.clj]: it's not easy to set up for that, we need to be able to define a not-null constraint for multiple columns, and it's currently not possible with our `defdataset`. 

it's good to cover that but I think it's too much effort.
I don't think we'll get that case a lot in the wild tho.

Comment (calherries) [reply test/metabase/api/action_test.clj]: I'm not sure it's that difficult to justify not covering it. We just need a table with composite PKs right? Here's an example.

https://github.com/metabase/metabase/blob/a47039c6b6d67aa811ae0d8e3f2c91195fe01e35/test/metabase/driver/mysql_test.clj#L448

You could copy+paste the code and change the json column to another type to make it easier.

File: test/metabase/driver/h2_test.clj [hunk 22]
```diff
@@ -343,3 +346,53 @@
...
+(deftest actions-maybe-parse-sql-error-test
+  (testing "violate not null constraint"
+    (is (= {:type    :metabase.actions.error/violate-not-null-constraint
+            :message "Ranking must have values."
+            :errors  {"RANKING" "You must provide a value."}}
+           (sql-jdbc.actions/maybe-parse-sql-error
+            :h2 actions.error/violate-not-null-constraint nil :row/created
+            "NULL not allowed for column \"RANKING\"; SQL statement:\nINSERT INTO \"PUBLIC\".\"GROUP\" (\"NAME\") VALUES (CAST(? AS VARCHAR)) [23502-214])"))))
+
+  (testing "vioalte unique constraint"
```
[the remark this answers is not in the document]
Comment (calherries) [reply test/metabase/driver/h2_test.clj]: bump

[the remark this answers is not in the document]
Comment (qnkhuat) [reply test/metabase/driver/h2_test.clj]: I fixed those in the new tests https://github.com/metabase/metabase/pull/32935/commits/cc95f072c399c1165df95478a655e597974483ee

File: test/metabase/driver/sql_jdbc/actions_test.clj [hunk 23]
```diff
@@ -51,3 +53,139 @@
...
+
+(defn- perform-action-ex-data
+  [& args]
+  (try
+   (apply actions/perform-action! args)
+   (catch Exception e
+     (ex-data e))))
+
+(deftest action-error-handling-test
+  (mt/test-drivers (mt/normal-drivers-with-feature :actions)
```
Comment (calherries) [inline test/metabase/driver/sql_jdbc/actions_test.clj:78]: I think this should be restricted to JDBC drivers for the future implementors of actions for non-jdbc drivers to skip these tests
```suggestion
  (mt/test-drivers (filter #(isa? driver/hierarchy % :sql-jdbc) (mt/normal-drivers-with-feature :actions))
```

**Comment by kamilmielnik** [comment]
I've merged #33258 here which is the frontend part of form-field-level error handling when executing actions + e2e tests for it.

**Review by calherries — APPROVED** [review]
LGTM, thanks for adding exhaustive tests
